### PR TITLE
Allow prefix/suffix options for store accessors

### PR DESCRIPTION
**Pull request by untidy-hair**
### Summary

####Background:
```ruby
class SampleModel < ApplicationRecord
  store :settings, accessors: [:color, :dependencies], coder: JSON
  has_many :dependencies
end
```
There was a situation where association ```has_many :dependencies``` existed
whereas we needed the same key "dependencies" (the value is array of dependency's names) as a part of "settings" JSON, too, so that the JSON can be consumed as it is from some other stuff.

We needed SampleModel#dependencies to return the association dependencies (and thus we wanted to add a method like setting_dependencies for JSON dependencies),
but the default behavior of SampleModel#dependencies above is to return JSON dependencies.

####Workarounds:
1. remove_method, undef or alias, etc, to override
 => Didn't quite work well.
2. ```has_many :my_dependencies, class: Dependency``` and have ```def dependencies; my_dependencies; end``` to override
 => Too much carefulness is necessary to cover other methods that has_many generates so I avoided this.
3. Do not apply ```store (or store_accessor)``` for JSON dependencies, and add ```def setting_dependencies; settings[:dependencies]; end``` (and setter also).
 => This was the most viable option. But dependencies does not show up in SampleModel.stored_attributes (however you can add it).

But in the end, among stored accessors, only dependencies had ```setting_``` prefix.
To be consistent and clearer, I ended up with adding ```setting_``` for other stored accessors, too, in the similar way as No3 above. (setting_color/setting_dependencies for SampleModel above.)

This PR allows this prefix as an option out of the box.
The above was just one case but I think there will be some other cases, too, where prefix/suffix option is convenient and useful.
(Btw I referenced enum prefix for the implementation.)

```ruby
store :settings, accessors: [ :two_factor_auth ], coder: JSON, _prefix: true
# => accessor will be model.settings_two_factor_auth
store_accessor :settings, :secret_question, _prefix: 'config'
# => accessor will be model.config_secret_question
store :settings, accessors: [ :login_retry ], _suffix: 'setting'
# => accessor will be model.login_retry_setting

```

Lastly, thank you for the great work, Rails team!

### CODE REVIEW

File: activerecord/lib/active_record/store.rb
Comment (gmcgibbon) [inline activerecord/lib/active_record/store.rb]: I think you can do `extract_options!` here instead.

File: activerecord/lib/active_record/store.rb
Comment (gmcgibbon) [inline activerecord/lib/active_record/store.rb]: Should it be `_prefix` or `prefix` like in active support's `delegate`? I guess `enum` have underscored options but I'm not sure which is preferred.

File: activerecord/lib/active_record/store.rb
Comment (gmcgibbon) [inline activerecord/lib/active_record/store.rb]: Should the examples be using symbols instead of strings for `_prefix` and `_suffix` values?

**Review by gmcgibbon — COMMENTED** [review]
Still relevant in `6.0.0.alpha`. I think this is a good idea! See comments for some suggestions 👍

**Comment by untidy-hair** [comment]
Thanks for the comments @gmcgibbon. Looks like another similar PR #32306 was created and merged before they had noticed this PR... @rafaelfranca (or someone), if you think it's worth adding suffix option or this PR can improve something related, I would be happy to contribute!

**Comment by gmcgibbon** [comment]
Ah, thanks for pointing that out @untidy-hair. I'd say you're good to rebase and add the suffix option but if you want to wait for a core member's approval that's cool too 👍

**Comment by untidy-hair** [comment]
well, i'm not waiting for an approval but i just want to make sure they recognize this PR this time haha :P

File: activerecord/lib/active_record/store.rb
Comment (rafaelfranca) [reply activerecord/lib/active_record/store.rb]: Without underscore, is the preffered way.

**Comment by untidy-hair** [comment]
I'm going to work on this this weekend :)

**Comment by untidy-hair** [comment]
Rebased and updated per comments :) @gmcgibbon, @rafaelfranca please review again. Thanks!

**Review by gmcgibbon — APPROVED** [review]
👍

**Comment by untidy-hair** [comment]
Thanks for the review @gmcgibbon! @rafaelfranca, can we merge this?

File: activerecord/lib/active_record/store.rb
Comment (kamipo) [inline activerecord/lib/active_record/store.rb]: Why do we define the helper methods in each `_store_accessors_module`s?

Comment (untidy-hair) [reply activerecord/lib/active_record/store.rb]: @kamipo, thanks for the review! The definition is there so that we can call the chunk of code in _store_accessors_module context, but it's true that each model that calls store/store_accessor defines the helper method inside the _store_accessors_module. Maybe I should just move that part of code directly into inside the module_eval block, unless you have any better suggestion. If you do, let me know :)

**Comment by untidy-hair** [comment]
@kamipo update. please review again. Thanks!

File: activerecord/lib/active_record/store.rb
Comment (kamipo) [inline activerecord/lib/active_record/store.rb]: Looks like that isn't correct appeared a subclass' accessor `:secret_question`.

File: activerecord/lib/active_record/store.rb
Comment (kamipo) [inline activerecord/lib/active_record/store.rb]:
```ruby
store_accessor(store_attribute, options[:accessors], options.slice(:prefix, :suffix)) if options.has_key? :accessors`
```

File: activerecord/lib/active_record/store.rb
Comment (kamipo) [inline activerecord/lib/active_record/store.rb]:
```ruby
def store_accessor(store_attribute, *keys, prefix: nil, suffix: nil)
```

File: activerecord/lib/active_record/store.rb
Comment (kamipo) [inline activerecord/lib/active_record/store.rb]: The prefix/suffix is just enough calculated once, there is no reason to do for each keys.

**Comment by untidy-hair** [comment]
@kamipo thanks for the comments. updated.

**Comment by kamipo** [comment]
Thanks!

**Pull request merged**
